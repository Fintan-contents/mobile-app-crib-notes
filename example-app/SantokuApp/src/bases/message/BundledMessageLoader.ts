import {bundledMessages, BundledMessagesType} from './BundledMessages';
import {MessagesLoader} from './Message';

/**
 * アプリ内にバンドルされたメッセージをロードするクラスです。
 */
class BundledMessagesLoader implements MessagesLoader {
  load() {
    return new Promise<BundledMessagesType>(resolve => resolve(bundledMessages));
  }
}

export {BundledMessagesLoader};
