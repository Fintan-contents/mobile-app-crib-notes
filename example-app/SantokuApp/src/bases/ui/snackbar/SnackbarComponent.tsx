import React, {useCallback, useRef, useState} from 'react';
import {Animated, StyleProp, StyleSheet, Text, TextStyle, TouchableOpacity, View, ViewStyle} from 'react-native';

import CompositeAnimation = Animated.CompositeAnimation;

export type SnackbarShowProps = {
  /**
   * Displayed message.
   */
  message: string;
  /**
   * Style of message.
   */
  messageTextStyle?: StyleProp<TextStyle>;
  /**
   * Style to adjust the look of the snackbar.
   */
  style?: StyleProp<ViewStyle>;
  /**
   * Adjust the position of the snackbar style.
   */
  positionStyle?: StyleProp<ViewStyle>;
  /**
   * Button text.
   * Placed to the right of the message.
   */
  actionText?: string;
  /**
   * Function called when the user taps the button.
   */
  actionHandler?: () => void;
  /**
   * Style of Button text.
   */
  actionTextStyle?: StyleProp<TextStyle>;
  /**
   * Time from when the snackbar appears to when it starts to fade-out.
   */
  autoHideDuration?: number;
  /**
   * Snackbar fade-in time.
   */
  fadeInDuration?: number;
  /**
   * Snackbar fade-out time.
   */
  fadeOutDuration?: number;
  /**
   * Fade-out time to close the previously displayed snack bar when displaying a new one.
   * This is applied when you try to display another snackbar while a snackbar is being displayed.
   */
  forceFadeOutDuration?: number;
  /**
   * Timestamp
   */
  timestamp?: number;
};

export type SnackbarHideProps = {
  /**
   * Specify when to hide the snackbar.
   */
  hide?: true;
  /**
   * Fade-out time when hide is specified.
   */
  hideFadeOutDuration?: number;
};

export type SnackbarProps = SnackbarShowProps & SnackbarHideProps;

export const SnackbarComponent: React.FC<SnackbarProps> = props => {
  const fadeAnim = useRef(new Animated.Value(0)).current;
  const fadeInAnimationRef = useRef<CompositeAnimation>();
  const fadeOutAnimationRef = useRef<CompositeAnimation>();
  const barrierFadeOutAnimationRef = useRef<CompositeAnimation>();
  const [visibleSnackbarProps, setVisibleSnackbarProps] = useState<SnackbarShowProps>();

  const animationStart = useCallback(
    (
      animationRef: React.MutableRefObject<CompositeAnimation | undefined>,
      config: Animated.TimingAnimationConfig,
      callback?: Animated.EndCallback,
    ) => {
      const animation = Animated.timing(fadeAnim, config);
      animationRef.current = animation;
      animation.start(result => {
        animationRef.current = undefined;
        if (callback) {
          callback(result);
        }
      });
    },
    [fadeAnim],
  );

  const show = useCallback(() => {
    setVisibleSnackbarProps({
      actionHandler: props.actionHandler,
      actionText: props.actionText,
      actionTextStyle: props.actionTextStyle,
      autoHideDuration: props.autoHideDuration,
      fadeInDuration: props.fadeInDuration,
      fadeOutDuration: props.fadeOutDuration,
      forceFadeOutDuration: props.forceFadeOutDuration,
      message: props.message,
      messageTextStyle: props.messageTextStyle,
      positionStyle: props.positionStyle,
      style: props.style,
      timestamp: props.timestamp,
    });

    const fadeInAnimationConfig = {
      toValue: 1,
      duration: props.fadeInDuration,
      useNativeDriver: true,
    };
    animationStart(fadeInAnimationRef, fadeInAnimationConfig, ({finished}) => {
      if (finished) {
        const fadeOutAnimationConfig = {
          toValue: 0,
          delay: props.autoHideDuration,
          duration: props.fadeOutDuration,
          useNativeDriver: true,
        };
        animationStart(fadeOutAnimationRef, fadeOutAnimationConfig, () => {
          if (!barrierFadeOutAnimationRef.current) {
            setVisibleSnackbarProps(undefined);
          }
        });
      } else {
        if (!barrierFadeOutAnimationRef.current) {
          setVisibleSnackbarProps(undefined);
        }
      }
    });
  }, [
    animationStart,
    props.actionHandler,
    props.actionText,
    props.actionTextStyle,
    props.autoHideDuration,
    props.fadeInDuration,
    props.fadeOutDuration,
    props.forceFadeOutDuration,
    props.message,
    props.messageTextStyle,
    props.positionStyle,
    props.style,
    props.timestamp,
  ]);

  const forceFadeout = useCallback(
    (fadeOutDuration?: number, callback?: () => void) => {
      const barrierFadeOutAnimationConfig = {
        toValue: 0,
        duration: fadeOutDuration,
        useNativeDriver: true,
      };

      barrierFadeOutAnimationRef.current?.stop();

      animationStart(barrierFadeOutAnimationRef, barrierFadeOutAnimationConfig, ({finished}) => {
        if (finished) {
          setVisibleSnackbarProps(undefined);
          callback?.();
        }
      });
      fadeInAnimationRef.current?.stop();
      fadeOutAnimationRef.current?.stop();
    },
    [animationStart],
  );

  React.useEffect(() => {
    if (props.hide) {
      forceFadeout(props.hideFadeOutDuration);
      return;
    }
    if (!props.message) {
      return;
    }
    if (barrierFadeOutAnimationRef.current) {
      forceFadeout(props.forceFadeOutDuration, show);
      return;
    }
    if (fadeInAnimationRef.current || fadeOutAnimationRef.current) {
      forceFadeout(props.forceFadeOutDuration, show);
      return;
    }
    show();
  }, [show, forceFadeout, props.hide, props.message, props.hideFadeOutDuration, props.forceFadeOutDuration]);

  const snackbarStyle = StyleSheet.flatten([styles.snackbar, props.style]);

  const animatedViewStyle = StyleSheet.flatten<ViewStyle>([styles.animatedContainer, props.positionStyle]);
  if (animatedViewStyle?.top === undefined && animatedViewStyle.bottom === undefined) {
    animatedViewStyle.bottom = 20;
  }

  const messageTextStyle = StyleSheet.flatten([styles.messageText, props.messageTextStyle]);
  const actionTextStyle = StyleSheet.flatten([styles.actionText, props.actionTextStyle]);

  if (!visibleSnackbarProps) {
    return null;
  }

  return (
    <Animated.View style={StyleSheet.flatten([{opacity: fadeAnim}, animatedViewStyle])} testID="snackbarAnimatedView">
      <View style={snackbarStyle} testID="snackbarStyleView">
        <View style={styles.messageContainer}>
          <Text style={messageTextStyle}>{visibleSnackbarProps.message}</Text>
        </View>
        {visibleSnackbarProps.actionText && visibleSnackbarProps.actionHandler && (
          <View style={styles.actionContainer}>
            <TouchableOpacity onPress={visibleSnackbarProps.actionHandler}>
              <Text style={actionTextStyle}>{visibleSnackbarProps.actionText}</Text>
            </TouchableOpacity>
          </View>
        )}
      </View>
    </Animated.View>
  );
};

SnackbarComponent.defaultProps = {
  autoHideDuration: 4000,
  fadeInDuration: 1000,
  fadeOutDuration: 1000,
  forceFadeOutDuration: 300,
  hideFadeOutDuration: 300,
};

const styles = StyleSheet.create({
  animatedContainer: {
    flex: 1,
    position: 'absolute',
    minHeight: 48,
    paddingHorizontal: 10,
    width: '100%',
  },
  snackbar: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    borderRadius: 5,
    backgroundColor: '#393939',
  },
  messageContainer: {
    flex: 1,
    paddingVertical: 20,
    paddingLeft: 20,
    paddingRight: 10,
  },
  actionContainer: {
    flex: 0,
    justifyContent: 'flex-end',
    alignItems: 'center',
    paddingVertical: 20,
    paddingLeft: 10,
    paddingRight: 20,
  },
  messageText: {
    color: 'white',
  },
  actionText: {
    color: '#85A3F1',
  },
});
