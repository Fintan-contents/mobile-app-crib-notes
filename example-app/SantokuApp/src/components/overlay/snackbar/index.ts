export * from './Snackbar';
export * from './WithSnackbar';
