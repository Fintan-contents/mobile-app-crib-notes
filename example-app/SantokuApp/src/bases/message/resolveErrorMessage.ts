import axios, {AxiosError} from 'axios';
import {m} from 'bases/message/Message';

let resolveErrorMessage = (error: unknown): {title: string; message: string} => {
  if (axios.isAxiosError(error)) {
    return resolveErrorMessageWhenAxiosError(error);
  }
  // 想定外のエラー
  return resolveErrorMessageWhenUnexpectedError();
};

function resolveErrorMessageWhenAxiosError(error: AxiosError) {
  const statusCode = error.response?.status;
  const title = m('システムエラー');
  const content = m('通信中にエラーが発生しました。');
  return {title, message: formatMessage(content, statusCode)};
}

function resolveErrorMessageWhenUnexpectedError() {
  const title = m('システムエラー');
  const content = m('しばらく経ってからもう1度お試しください。');
  return {title, message: formatMessage(content)};
}

function formatMessage(content: string, statusCode?: number) {
  if (statusCode === undefined) {
    return `${content}`;
  }
  return `${content}\n${statusCode}`;
}

const setErrorMessageResolver = (e: (error: unknown) => {title: string; message: string}) => {
  resolveErrorMessage = e;
};

export {resolveErrorMessage, setErrorMessageResolver};
