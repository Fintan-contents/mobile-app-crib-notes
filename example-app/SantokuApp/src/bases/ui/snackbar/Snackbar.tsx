import {m} from 'bases/message/Message';
import React, {useMemo, useState} from 'react';

import {
  SnackbarComponent,
  SnackbarHideProps as SnackbarComponentHideProps,
  SnackbarProps as SnackbarComponentProps,
  SnackbarShowProps as SnackbarComponentShowProps,
} from './SnackbarComponent';
import {FullWindowOverlay} from '../overlay/FullWindowOverlay';

type SnackbarShowProps = Omit<SnackbarComponentShowProps, 'message'>;
type SnackbarShowCloseButtonProps = Omit<SnackbarShowProps, 'actionText' | 'actionHandler' | 'actionTextStyle'>;
type SnackbarHideProps = Omit<SnackbarComponentHideProps, 'hide'>;

type SnackbarType = {
  /**
   * Show snackbar.
   *
   * @param message - Displayed message.
   * @param showProps - Snackbar props(SnackbarShowContextProps).
   */
  show: (message: string, showProps?: SnackbarShowProps) => void;

  /**
   * Show the snackbar with the close button.
   * The close button is placed to the right of the message.
   *
   * @param message - Displayed message.
   * @param showProps - Snackbar props(SnackbarShowCloseButtonContextProps).
   */
  showWithCloseButton: (message: string, showProps?: SnackbarShowCloseButtonProps) => void;

  /**
   * Hide snackbar.
   *
   * @param hideProps - Snackbar props(SnackbarHideContextProps).
   */
  hide: (hideProps?: SnackbarHideProps) => void;
  Component: typeof Component;
};

let Snackbar: SnackbarType = {
  show: () => {
    throw new Error('Snackbar.Component is required.');
  },
  showWithCloseButton: () => {
    throw new Error('Snackbar.Component is required.');
  },
  hide: () => {
    throw new Error('Snackbar.Component is required.');
  },
  Component,
};

function Component(props: {initialState?: SnackbarComponentShowProps}) {
  const [state, setState] = useState<SnackbarComponentProps>(
    props.initialState ?? {
      message: '',
    },
  );
  // const snackbarContext = useMemo<SnackbarContextType>(
  Snackbar = useMemo<SnackbarType>(
    () => ({
      ...Snackbar,
      show: (message: string, showProps?: SnackbarShowProps) => {
        setState({
          timestamp: Date.now(),
          ...showProps,
          message,
        });
      },
      showWithCloseButton: (message: string, showProps?: SnackbarShowCloseButtonProps) => {
        setState({
          timestamp: Date.now(),
          ...showProps,
          actionText: m('閉じる'),
          actionHandler: Snackbar.hide,
          message,
        });
      },
      hide: (hideProps?: SnackbarHideProps) => {
        setState(prevState => ({...prevState, ...hideProps, hide: true}));
      },
    }),
    [],
  );
  return (
    <>
      <FullWindowOverlay>
        <SnackbarComponent {...state} />
      </FullWindowOverlay>
    </>
  );
}

export {Snackbar};
