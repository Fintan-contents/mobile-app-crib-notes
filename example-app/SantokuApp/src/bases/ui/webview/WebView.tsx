import {m} from 'bases/message/Message';
import React, {useCallback, useEffect, useState} from 'react';
import {ActivityIndicator, StyleSheet} from 'react-native';
import {WebView as RNWebView, WebViewProps} from 'react-native-webview';
import {
  WebViewErrorEvent,
  WebViewNavigationEvent,
  WebViewProgressEvent,
  WebViewScrollEvent,
  WebViewSource,
  WebViewSourceUri,
} from 'react-native-webview/lib/WebViewTypes';

import {Snackbar} from '../snackbar/Snackbar';

type Props = WebViewProps & {
  /**
   * Error message when React Native WebView original onError is raised.
   * If not specified, the default message is displayed.
   * If onError is specified by the parent, this message will not be displayed.
   */
  errorMessage?: string;
  /**
   * End-of-Scroll Event.
   * Occurs every time the web page scrolls to the bottom of the content.
   */
  onScrollEnd?: () => void;
  /**
   * End-of-scroll event that occurs only once.
   * Occurs only once when the web page scroll reaches the bottom of the content.
   */
  onScrollEndOnce?: () => void;
};

const isUriSource = (source?: WebViewSource): source is WebViewSourceUri => {
  return source !== undefined && 'uri' in source;
};

/**
 * WebViewでページを表示するコンポーネント。
 * react-native-webviewで提供されるイベントに、以下の2つのイベントを追加しています。
 * - onScrollEnd: スクロールがページの終端に移動する度に発生
 * - onScrollEndOnce: スクロールがページの終端に移動した最初の1回のみ発生
 *
 * 動作確認中に発生した以下の事象に対応しています。
 * - 発生OS
 *   - iOS
 * - 発生事象
 *   - URIが変わった直後にonLoadStart->onLoadEnd->onScroll->onScrollの順で発火し、1回目のonScrollで渡されるcontentOffsetが、URIが変わる前のページのcontentOffsetになっている。
 *     そのため、URIが変わる前のページで終端までスクロールしていた場合、1回目のonScrollで終端までスクロールされていると判定されてしまい、onScrollEnd/onScrollOnceイベントが発生してしまう。
 *     なお、2回目のonScrollで渡されるcontentOffsetは0になっている
 * - 対応方法
 *   - 1回目のonScrollでonScrollEnd/onScrollEndOnceイベントが発生しないように、URIが変わったかどうかを状態として持ち、
 *     URIが変わった直後の、オフセットが0ではないスクロールイベントでは、onScrollEnd/onScrollEndOnceは発生させない。
 *
 * - 発生OS
 *   - Android
 * - 発生事象
 *   - onLoadEndがページのロードが完了する前に呼び出される不具合がある。（https://github.com/react-native-webview/react-native-webview/issues/2345）
 * - 対応方法
 *   - onLoadProgressイベント発生時に、progressが1になったらページロードが完了したと判定する。
 *
 * なお、本コンポーネントには以下の制約事項があります。
 * - 対象OS
 *   - Android
 * - 制約事項
 *   - スクロールが存在しないページでは、onScrollEnd/onScrollEndOnceイベントは発生しません。
 *
 */
export const WebView = React.forwardRef<RNWebView, Props>(function WebView(props, ref) {
  const [loadEnd, setLoadEnd] = useState(false);
  const [scrollEndCalled, setScrollEndCalled] = useState(false);
  const {onScrollEnd, onScrollEndOnce, onLoadStart, onLoadProgress, onError, errorMessage, ...webViewProps} = props;

  const [isUriChanged, setIsUriChanged] = useState(false);
  const uri = isUriSource(props.source) ? props.source.uri : undefined;
  useEffect(() => {
    setIsUriChanged(true);
  }, [uri]);

  // uriが変わった時にWebViewがページを再ロードした場合の対応
  // なお、WebViewで表示しているページ内リンクでページ遷移した場合は、onLoadStartイベントは再発行されません
  const handleLoadStart = useCallback(
    (event: WebViewNavigationEvent) => {
      setLoadEnd(false);
      setScrollEndCalled(false);
      onLoadStart?.(event);
    },
    [onLoadStart],
  );

  const handleScroll = useCallback(
    (event: WebViewScrollEvent) => {
      if (isUriChanged && event.nativeEvent.contentOffset.y > 0) {
        // URLが変わった直後の、オフセットが0ではないスクロールイベントでは、onScrollEndは発生させない。
        setIsUriChanged(false);
      } else if ((onScrollEnd || onScrollEndOnce) && loadEnd) {
        // 小数点の誤差があるため、1px分は丸め誤差として扱う
        const scrollY = event.nativeEvent.contentOffset.y + event.nativeEvent.layoutMeasurement.height + 1;
        if (event.nativeEvent.contentSize.height <= scrollY) {
          if (!scrollEndCalled) {
            setScrollEndCalled(true);
            onScrollEndOnce?.();
          }
          onScrollEnd?.();
        }
      }

      props.onScroll?.(event);
    },
    [isUriChanged, loadEnd, scrollEndCalled, onScrollEnd, onScrollEndOnce, props],
  );

  const handleLoadProgress = useCallback(
    (event: WebViewProgressEvent) => {
      if (event.nativeEvent.progress === 1) {
        setLoadEnd(true);
      } else {
        setLoadEnd(false);
      }
      onLoadProgress?.(event);
    },
    [onLoadProgress],
  );

  const handleError = useCallback(
    (event: WebViewErrorEvent) => {
      if (onError) {
        onError(event);
      } else {
        Snackbar.showWithCloseButton(errorMessage ?? m('app.webview.onError'));
      }
    },
    [errorMessage, onError],
  );

  return (
    <RNWebView
      {...webViewProps}
      style={styles.container}
      onScroll={handleScroll}
      onLoadStart={handleLoadStart}
      onLoadProgress={handleLoadProgress}
      onError={handleError}
      ref={ref}
    />
  );
});

WebView.defaultProps = {
  // androidLayerType：WebViewの再レンダー時にクラッシュする可能性がある問題の回避策
  // https://github.com/react-native-webview/react-native-webview/issues/1915
  androidLayerType: 'software',
  startInLoadingState: true,
  // iOSでスクロールしたときに滑らかにスクロールするようにしておく。
  // https://github.com/react-native-webview/react-native-webview/blob/master/docs/Reference.md#decelerationrate
  decelerationRate: 'normal',
  renderLoading: () => <ActivityIndicator style={styles.indicator} size="large" color="#0000ff" />,
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  indicator: {
    position: 'absolute',
    height: '100%',
    width: '100%',
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
  },
});
