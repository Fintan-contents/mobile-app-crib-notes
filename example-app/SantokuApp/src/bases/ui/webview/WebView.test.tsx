import {fireEvent, render, screen} from '@testing-library/react-native';
import {BundledMessagesLoader} from 'bases/message/BundledMessageLoader';
import {loadMessages} from 'bases/message/Message';
import React from 'react';

import {WebView} from './WebView';
import {Snackbar} from '../snackbar/Snackbar';

describe('WebView', () => {
  it('WebViewが正常にrenderできることを確認', () => {
    render(<WebView testID="webview" source={{uri: 'https://localhost/'}} />);
    expect(screen.queryByTestId('webview')).not.toBeNull();
    expect(screen).toMatchSnapshot();
  });

  it('WebViewのスクロールイベントが適切なタイミングで発行されることを確認', () => {
    const handleOnLoadStart = jest.fn();
    const handleOnScroll = jest.fn();
    const handleOnScrollEnd = jest.fn();
    const handleOnScrollEndOnce = jest.fn();

    const layoutMeasurement = {height: 600};
    const contentSize = {height: 2000};

    render(
      <WebView
        source={{uri: 'https://localhost/'}}
        onLoadStart={handleOnLoadStart}
        onScroll={handleOnScroll}
        onScrollEnd={handleOnScrollEnd}
        onScrollEndOnce={handleOnScrollEndOnce}
        testID="webview"
      />,
    );

    const webview = screen.getByTestId('webview');

    expect(webview).not.toBeNull();

    // onLoadStartイベント発生前
    expect(handleOnLoadStart).not.toHaveBeenCalled();
    fireEvent(webview, 'onLoadStart');
    // onLoadStartイベント発生後
    expect(handleOnLoadStart).toHaveBeenCalled();

    // ロード完了前はonScrollEndイベントを発行しない
    fireEvent.scroll(webview, {nativeEvent: {contentOffset: {y: 1400}, layoutMeasurement, contentSize}});
    expect(handleOnScroll).toHaveBeenCalledTimes(1);
    expect(handleOnScrollEnd).not.toHaveBeenCalled();
    expect(handleOnScrollEndOnce).not.toHaveBeenCalled();

    // onLoadProgressイベントをprogress:1で発生させる
    fireEvent(webview, 'onLoadProgress', {nativeEvent: {progress: 1}});

    // スクロール終端判定に対して、0.1 未達
    fireEvent.scroll(webview, {nativeEvent: {contentOffset: {y: 1398.9}, layoutMeasurement, contentSize}});
    expect(handleOnScroll).toHaveBeenCalledTimes(2);
    expect(handleOnScrollEnd).not.toHaveBeenCalled();
    expect(handleOnScrollEndOnce).not.toHaveBeenCalled();

    // スクロール終端と判定（１回目）
    fireEvent.scroll(webview, {nativeEvent: {contentOffset: {y: 1399}, layoutMeasurement, contentSize}});
    expect(handleOnScroll).toHaveBeenCalledTimes(3);
    expect(handleOnScrollEnd).toHaveBeenCalledTimes(1);
    expect(handleOnScrollEndOnce).toHaveBeenCalledTimes(1);

    // スクロール終端と判定（２回目のため、onScrollEndOnceイベントは発生しない）
    fireEvent.scroll(webview, {nativeEvent: {contentOffset: {y: 1400}, layoutMeasurement, contentSize}});
    expect(handleOnScroll).toHaveBeenCalledTimes(4);
    expect(handleOnScrollEnd).toHaveBeenCalledTimes(2);
    expect(handleOnScrollEndOnce).toHaveBeenCalledTimes(1);

    screen.update(
      <WebView
        source={{uri: 'https://localhost2/'}}
        onLoadStart={handleOnLoadStart}
        onScroll={handleOnScroll}
        onScrollEnd={handleOnScrollEnd}
        onScrollEndOnce={handleOnScrollEndOnce}
        testID="webview"
      />,
    );
    // uriが変わって再度onLoadStartイベントを発生させる
    fireEvent(webview, 'onLoadStart');
    // ロード完了前にスクロールイベントを発生させる
    fireEvent.scroll(webview, {nativeEvent: {contentOffset: {y: 1399}, layoutMeasurement, contentSize}});
    // ロード完了前はonScrollEndイベントが発生しないこと
    expect(handleOnScrollEnd).toHaveBeenCalledTimes(2);
    // onLoadProgressイベントをprogress:1で発生させる
    fireEvent(webview, 'onLoadProgress', {nativeEvent: {progress: 1}});
    // ロード完了後にスクロール終端と判定
    fireEvent.scroll(webview, {nativeEvent: {contentOffset: {y: 1399}, layoutMeasurement, contentSize}});
    // onScrollEndイベントが発生
    expect(handleOnScrollEnd).toHaveBeenCalledTimes(3);
    // ページが再ロードされたため、onScrollEndOnceイベントが発生
    expect(handleOnScrollEndOnce).toHaveBeenCalledTimes(2);

    screen.update(
      <WebView
        source={{uri: 'https://localhost3/'}}
        onLoadStart={handleOnLoadStart}
        onScroll={handleOnScroll}
        onScrollEnd={handleOnScrollEnd}
        onScrollEndOnce={handleOnScrollEndOnce}
        testID="webview"
      />,
    );
    // uriが変わって再度onLoadStartイベントを発生させる
    fireEvent(webview, 'onLoadStart');
    // onLoadProgressイベントをprogress:1で発生させる
    fireEvent(webview, 'onLoadProgress', {nativeEvent: {progress: 1}});
    // ロード完了後に、末尾までスクロールするイベントを発生させる
    fireEvent.scroll(webview, {nativeEvent: {contentOffset: {y: 1400}, layoutMeasurement, contentSize}});
    // URL変更直後の、オフセット0以外のスクロールイベントはonScrollEndイベントを発生させないこと。
    expect(handleOnScrollEnd).toHaveBeenCalledTimes(3);
    // もう一度スクロールイベントを発生させる。
    fireEvent.scroll(webview, {nativeEvent: {contentOffset: {y: 1400}, layoutMeasurement, contentSize}});
    // onScrollEndイベントが発生
    expect(handleOnScrollEnd).toHaveBeenCalledTimes(4);

    const contentSizeShort = {height: 200};
    screen.update(
      <WebView
        source={{uri: 'https://localhost4/'}}
        onLoadStart={handleOnLoadStart}
        onScroll={handleOnScroll}
        onScrollEnd={handleOnScrollEnd}
        onScrollEndOnce={handleOnScrollEndOnce}
        testID="webview"
      />,
    );
    // uriが変わって再度onLoadStartを発生させる
    fireEvent(webview, 'onLoadStart');
    // onLoadProgressイベントをprogress:1で発生させる
    fireEvent(webview, 'onLoadProgress', {nativeEvent: {progress: 1}});
    // ロード完了後に、オフセット0のスクロールするイベントを発生させる
    fireEvent.scroll(webview, {nativeEvent: {contentOffset: {y: 0}, layoutMeasurement, contentSize: contentSizeShort}});
    // URL変更直後でも、オフセット0のスクロールイベントで終端まで表示されていればonScrollEndイベントが発生する
    expect(handleOnScrollEnd).toHaveBeenCalledTimes(5);
  });

  it('WebViewにonScrollEndのみ指定して正常にイベントが発行されることを確認', () => {
    const handleOnScrollEnd = jest.fn();

    const layoutMeasurement = {height: 600};
    const contentSize = {height: 2000};

    render(<WebView source={{uri: 'https://localhost/'}} onScrollEnd={handleOnScrollEnd} testID="webview" />);
    const webview = screen.getByTestId('webview');

    // ロード完了前のスクロールイベント
    fireEvent.scroll(webview, {nativeEvent: {contentOffset: {y: 1400}, layoutMeasurement, contentSize}});
    expect(handleOnScrollEnd).not.toHaveBeenCalled();

    // onLoadProgressイベントをprogress:1で発生させる
    fireEvent(webview, 'onLoadProgress', {nativeEvent: {progress: 1}});

    // スクロール終端判定に対して、0.1 未達
    fireEvent.scroll(webview, {nativeEvent: {contentOffset: {y: 1398.9}, layoutMeasurement, contentSize}});
    expect(handleOnScrollEnd).not.toHaveBeenCalled();

    // スクロール終端と判定
    fireEvent.scroll(webview, {nativeEvent: {contentOffset: {y: 1399}, layoutMeasurement, contentSize}});
    expect(handleOnScrollEnd).toHaveBeenCalledTimes(1);
  });

  it('WebViewにonScrollEndOnceのみ指定して正常にイベントが発行されることを確認', () => {
    const handleOnScrollEndOnce = jest.fn();

    const layoutMeasurement = {height: 600};
    const contentSize = {height: 2000};

    render(<WebView source={{uri: 'https://localhost/'}} onScrollEndOnce={handleOnScrollEndOnce} testID="webview" />);
    const webview = screen.getByTestId('webview');

    // ロード完了前のスクロールイベント
    fireEvent.scroll(webview, {nativeEvent: {contentOffset: {y: 1400}, layoutMeasurement, contentSize}});
    expect(handleOnScrollEndOnce).not.toHaveBeenCalled();

    // onLoadProgressイベントをprogress:1で発生させる
    fireEvent(webview, 'onLoadProgress', {nativeEvent: {progress: 1}});

    // スクロール終端判定に対して、0.1 未達
    fireEvent.scroll(webview, {nativeEvent: {contentOffset: {y: 1398.9}, layoutMeasurement, contentSize}});
    expect(handleOnScrollEndOnce).not.toHaveBeenCalled();

    // スクロール終端と判定
    fireEvent.scroll(webview, {nativeEvent: {contentOffset: {y: 1399}, layoutMeasurement, contentSize}});
    expect(handleOnScrollEndOnce).toHaveBeenCalledTimes(1);
  });

  it('WebViewにonScrollのみ指定して正常にイベントが発行されることを確認', () => {
    const handleOnScroll = jest.fn();

    const layoutMeasurement = {height: 600};
    const contentSize = {height: 2000};

    render(<WebView source={{uri: 'https://localhost/'}} onScroll={handleOnScroll} testID="webview" />);
    const webview = screen.getByTestId('webview');

    // ロード完了前のスクロールイベント
    fireEvent.scroll(webview, {nativeEvent: {contentOffset: {y: 1400}, layoutMeasurement, contentSize}});
    expect(handleOnScroll).toHaveBeenCalledTimes(1);

    // onLoadProgressイベントをprogress:1で発生させる
    fireEvent(webview, 'onLoadProgress', {nativeEvent: {progress: 1}});

    // スクロール終端判定に対して、0.1 未達
    fireEvent.scroll(webview, {nativeEvent: {contentOffset: {y: 1398.9}, layoutMeasurement, contentSize}});
    expect(handleOnScroll).toHaveBeenCalledTimes(2);

    // スクロール終端と判定
    fireEvent.scroll(webview, {nativeEvent: {contentOffset: {y: 1399}, layoutMeasurement, contentSize}});
    expect(handleOnScroll).toHaveBeenCalledTimes(3);
  });

  it('WebViewのonErrorイベントで、親からonErrorを指定してない場合はSnackbar表示関数をコールすることを確認', async () => {
    const spySnackbar = jest.spyOn(Snackbar, 'showWithCloseButton').mockImplementation(() => {});
    await loadMessages(new BundledMessagesLoader());

    render(<WebView source={{uri: 'https://localhost/'}} testID="webview" />);
    const webview = screen.getByTestId('webview');

    fireEvent(webview, 'onError');
    expect(spySnackbar).toHaveBeenCalledTimes(1);
  });

  it('WebViewのonErrorイベントで、親からonErrorを指定した場合は親へイベント通知することを確認', () => {
    const handleOnError = jest.fn();

    render(<WebView source={{uri: 'https://localhost/'}} onError={handleOnError} testID="webview" />);
    const webview = screen.getByTestId('webview');

    fireEvent(webview, 'onError');
    expect(handleOnError).toHaveBeenCalledTimes(1);
  });

  it('WebViewSourceHtmlでも正常にレンダリングされること', () => {
    const handleOnError = jest.fn();

    render(<WebView source={{html: '<h1>test</h1>'}} onError={handleOnError} testID="webview" />);
    expect(screen.queryByTestId('webview')).not.toBeNull();
    expect(screen).toMatchSnapshot();
  });
});
