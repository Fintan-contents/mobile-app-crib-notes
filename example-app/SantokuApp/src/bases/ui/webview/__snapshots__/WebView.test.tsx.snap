// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebView WebViewSourceHtmlでも正常にレンダリングされること 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "overflow": "hidden",
      },
      undefined,
    ]
  }
>
  <RNCWebView
    androidLayerType="software"
    cacheEnabled={true}
    decelerationRate={0.998}
    injectedJavaScriptBeforeContentLoadedForMainFrameOnly={true}
    injectedJavaScriptForMainFrameOnly={true}
    javaScriptEnabled={true}
    messagingEnabled={true}
    onContentProcessDidTerminate={[Function]}
    onHttpError={[Function]}
    onLoadingError={[Function]}
    onLoadingFinish={[Function]}
    onLoadingProgress={[Function]}
    onLoadingStart={[Function]}
    onMessage={[Function]}
    onScroll={[Function]}
    onShouldStartLoadWithRequest={[Function]}
    source={
      Object {
        "html": "<h1>test</h1>",
      }
    }
    style={
      Array [
        Object {
          "flex": 1,
          "overflow": "hidden",
        },
        Object {
          "backgroundColor": "#ffffff",
        },
        Object {
          "flex": 1,
        },
      ]
    }
    testID="webview"
    textInteractionEnabled={true}
    useSharedProcessPool={true}
  />
  <ActivityIndicator
    color="#0000ff"
    size="large"
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "height": "100%",
        "justifyContent": "center",
        "position": "absolute",
        "width": "100%",
      }
    }
  />
</View>
`;

exports[`WebView WebViewが正常にrenderできることを確認 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "overflow": "hidden",
      },
      undefined,
    ]
  }
>
  <RNCWebView
    androidLayerType="software"
    cacheEnabled={true}
    decelerationRate={0.998}
    injectedJavaScriptBeforeContentLoadedForMainFrameOnly={true}
    injectedJavaScriptForMainFrameOnly={true}
    javaScriptEnabled={true}
    messagingEnabled={true}
    onContentProcessDidTerminate={[Function]}
    onHttpError={[Function]}
    onLoadingError={[Function]}
    onLoadingFinish={[Function]}
    onLoadingProgress={[Function]}
    onLoadingStart={[Function]}
    onMessage={[Function]}
    onScroll={[Function]}
    onShouldStartLoadWithRequest={[Function]}
    source={
      Object {
        "uri": "https://localhost/",
      }
    }
    style={
      Array [
        Object {
          "flex": 1,
          "overflow": "hidden",
        },
        Object {
          "backgroundColor": "#ffffff",
        },
        Object {
          "flex": 1,
        },
      ]
    }
    testID="webview"
    textInteractionEnabled={true}
    useSharedProcessPool={true}
  />
  <ActivityIndicator
    color="#0000ff"
    size="large"
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "height": "100%",
        "justifyContent": "center",
        "position": "absolute",
        "width": "100%",
      }
    }
  />
</View>
`;
